intel-sgx/dcap-retrieve-pckid/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![deny(warnings)]
+use std::fmt;
+use std::convert::TryInto;
+
+use aesm_client::AesmClient;
+use dcap_ql::quote::{Qe3CertDataPpid, Quote, Quote3SignatureEcdsaP256, QuoteHeader};
+use sgx_isa::Targetinfo;
+#[cfg(windows)]
+use sgxs_loaders::enclaveapi::Sgx as IsgxDevice;
+#[cfg(unix)]
+use sgxs_loaders::isgx::Device as IsgxDevice;
+
+fn get_algorithm_id(key_id: &Vec<u8>) -> u32 {
+    const ALGORITHM_OFFSET: usize = 154;
+
+    let mut bytes: [u8; 4] = Default::default();
+    bytes.copy_from_slice(&key_id[ALGORITHM_OFFSET..ALGORITHM_OFFSET + 4]);
+    u32::from_le_bytes(bytes)
+}
+
+struct PrintHex<'a>(&'a [u8]);
+
+impl<'a> fmt::Display for PrintHex<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for b in self.0 {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+pub struct PckId {
+    pub cd_ppid: Qe3CertDataPpid<'static>,
```

Comment:
This may be me not understanding lifetimes, but do we really want to force the `Qe3CertDataPpid` field to have `'static` lifetime? Would it not be better to do something like this instead:
```
pub struct PckId<'a> {
    pub cd_ppid: Qe3CertDataPpid<'a>,
    pub qe_id: [u8; 16],
}
```

Reply:
Since, the `cd_ppid` is local variable created inside the function, so it needs to be a owned object to return.
After calling following function, the reference inside it is converted to `static`.

https://github.com/fortanix/rust-sgx/blob/f10c25b72081addc4843c2d779ef9660ff7cf60f/intel-sgx/dcap-retrieve-pckid/src/lib.rs#L97